File: backend/fasting-service/src/test/java/com/ohchurus/fasting/util/PeriodUtilsTest.java
package com.ohchurus.fasting.util;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.time.LocalDate;

import static org.junit.jupiter.api.Assertions.*;

@DisplayName("PeriodUtils (fasting)")
class PeriodUtilsTest {

    @Nested
    @DisplayName("getStartOfPeriod")
    class StartTests {

        @Test
        @DisplayName("Should return same-month start when reference is on/after the start day")
        void shouldReturnSameMonth() {
            LocalDate ref = LocalDate.of(2026, 3, 20);
            assertEquals(LocalDate.of(2026, 3, 15), PeriodUtils.getStartOfPeriod(15, ref));
        }

        @Test
        @DisplayName("Should roll back to previous month when reference is before the start day")
        void shouldReturnPreviousMonth() {
            LocalDate ref = LocalDate.of(2026, 3, 10);
            assertEquals(LocalDate.of(2026, 2, 15), PeriodUtils.getStartOfPeriod(15, ref));
        }

        @Test
        @DisplayName("Should clamp day 31 to last day on short months")
        void shouldClampShortMonth() {
            // February 2026 (non-leap, 28 days) with start day 31 -> 28
            LocalDate ref = LocalDate.of(2026, 2, 28);
            assertEquals(LocalDate.of(2026, 2, 28), PeriodUtils.getStartOfPeriod(31, ref));
        }
    }

    @Nested
    @DisplayName("getEndOfPeriod")
    class EndTests {

        @Test
        @DisplayName("Should end the day before next period start")
        void shouldComputeEnd() {
            LocalDate start = LocalDate.of(2026, 3, 15);
            assertEquals(LocalDate.of(2026, 4, 14), PeriodUtils.getEndOfPeriod(15, start));
        }

        @Test
        @DisplayName("Should clamp end when next month is shorter")
        void shouldClampEnd() {
            // start day 31 January -> next month Feb clamps to 28 -> end = 27 Feb
            LocalDate start = LocalDate.of(2026, 1, 31);
            assertEquals(LocalDate.of(2026, 2, 27), PeriodUtils.getEndOfPeriod(31, start));
        }
    }
}

File: frontend/src/utils/__tests__/periodUtils.test.ts
import { getStartOfPeriod, getEndOfPeriod, navigatePeriod } from '../periodUtils';

describe('periodUtils', () => {
  describe('getStartOfPeriod', () => {
    it('returns same-month start when reference is on/after the start day', () => {
      expect(getStartOfPeriod(15, new Date(2026, 2, 20))).toBe('2026-03-15');
    });

    it('rolls back to previous month when reference is before the start day', () => {
      expect(getStartOfPeriod(15, new Date(2026, 2, 10))).toBe('2026-02-15');
    });

    it('handles January reference rolling back to previous December', () => {
      expect(getStartOfPeriod(15, new Date(2026, 0, 5))).toBe('2025-12-15');
    });

    it('clamps the start day on short months', () => {
      // budgetStartDay 31 in February (28 days) -> clamp to 28
      expect(getStartOfPeriod(31, new Date(2026, 1, 28))).toBe('2026-02-28');
    });
  });

  describe('getEndOfPeriod', () => {
    it('ends the day before next period start', () => {
      expect(getEndOfPeriod(15, '2026-03-15')).toBe('2026-04-14');
    });

    it('handles December rolling into next year', () => {
      expect(getEndOfPeriod(1, '2026-12-01')).toBe('2026-12-31');
    });

    it('clamps when the next month is shorter', () => {
      expect(getEndOfPeriod(31, '2026-01-31')).toBe('2026-02-27');
    });
  });

  describe('navigatePeriod', () => {
    it('navigates to the next period', () => {
      expect(navigatePeriod('2026-03-15', 15, 'next')).toEqual({ start: '2026-04-15', end: '2026-05-14' });
    });

    it('navigates to the previous period', () => {
      expect(navigatePeriod('2026-03-15', 15, 'prev')).toEqual({ start: '2026-02-15', end: '2026-03-14' });
    });

    it('navigates next across a year boundary', () => {
      expect(navigatePeriod('2026-12-15', 15, 'next')).toEqual({ start: '2027-01-15', end: '2027-02-14' });
    });
  });
});
